source/cpu/pstate.go

```diff
@@ -32,12 +35,68 @@ func detectPstate() (map[string]string, error) {
 		return nil, nil
 	}
 
-	// Only looking for turbo boost for now...
+	// Get global pstate status
+	data, err := ioutil.ReadFile(source.SysfsDir.Path("devices/system/cpu/intel_pstate/status"))
+	if err != nil {
+		return nil, fmt.Errorf("could not read pstate status: %s", err.Error())
+	}
+	status := strings.TrimSpace(string(data))
+
+	// Check that intel_pstate driver is being used by looping through policies
+	policies, err := ioutil.ReadDir(source.SysfsDir.Path("devices/system/cpu/cpufreq"))
+	if err != nil {
+		return nil, fmt.Errorf("cpufreq directory not available to check pstate: %s", err.Error())
+	}
+
+	features := map[string]string{}
+	for _, policy := range policies {
+		// Ensure at least one cpu is using this policy
+		cpus, err := ioutil.ReadFile(source.SysfsDir.Path("devices/system/cpu/cpufreq", policy.Name(), "affected_cpus"))
+		if err != nil || strings.TrimSpace(string(cpus)) == "" {
+			klog.Errorf("cpufreq policy %s has no affected cpus", policy.Name())
+			continue
+		}
+
+		// Check if policy is using intel_pstate scaling driver
+		driver, err := ioutil.ReadFile(source.SysfsDir.Path("devices/system/cpu/cpufreq", policy.Name(), "scaling_driver"))
+		if err != nil {
+			klog.Errorf("cpufreq scaling driver cannot be read: %s", err.Error())
+			continue
+		}
+
+		if strings.TrimSpace(string(driver)) != "intel_pstate" {
+			continue
+		}
+
+		if status == "active" || status == "passive" {
+			features["status"] = status
+		}
+
+		if status == "active" {
+			// Check powersave/performance setting
+			data, err := ioutil.ReadFile(source.SysfsDir.Path("devices/system/cpu/cpufreq", policy.Name(), "scaling_governor"))
+			scaling := strings.TrimSpace(string(data))
+			if err == nil {
+				if scaling == "powersave" || scaling == "performance" {
+					features["policy"] = scaling
+				}
+			}
```

Comment:
I think we need to check that all policies share the same governor and not just take what happens to be the governor of the last policy

Reply:
OK, changed to check all policies and only set feature if all polices have the same scaling_governor.